pkg/controllers/cloudmap_controller.go

```diff
@@ -208,37 +208,114 @@ func (r *CloudMapReconciler) createAndGetDerivedService(ctx context.Context, svc
 	return r.getDerivedService(ctx, svc.Namespace, svcImport.Annotations[DerivedServiceAnnotation])
 }
 
-func (r *CloudMapReconciler) updateEndpointSlices(ctx context.Context, svcImport *v1alpha1.ServiceImport, cloudMapService *model.Service, svc *v1.Service) error {
+func (r *CloudMapReconciler) updateEndpointSlices(ctx context.Context, svcImport *v1alpha1.ServiceImport, desiredEndpoints []*model.Endpoint, svc *v1.Service) error {
 	existingSlicesList := discovery.EndpointSliceList{}
-	var existingSlices []*discovery.EndpointSlice
 	if err := r.Client.List(ctx, &existingSlicesList,
 		client.InNamespace(svc.Namespace), client.MatchingLabels{discovery.LabelServiceName: svc.Name}); err != nil {
 		return err
 	}
-	if len(existingSlicesList.Items) == 0 {
-		// create new endpoint slice
-		existingSlices = createEndpointSlicesStruct(svcImport, cloudMapService, svc)
-		for _, slice := range existingSlices {
-			if err := r.Client.Create(ctx, slice); err != nil {
+
+	desiredPorts := extractEndpointPorts(desiredEndpoints)
+	matchedEndpoints := make(map[string]*discovery.Endpoint)
+	endpointsToCreate := make([]discovery.Endpoint, 0)
+
+	// check if all desired endpoints are in an endpoint slice already
+	for _, desiredEndpoint := range desiredEndpoints {
+		existingEndpoint := findEndpointInSliceByIp(existingSlicesList.Items, desiredEndpoint.IP)
```

Comment:
This is O(n^2) operation. We can improve here by normalizing `existingSlicesList.Items` as `map[Address]Endpoint` before the for loop, and doing a key lookup.

Reply:
It's a good idea.